// express-csp-header/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const crypto_1 = __importDefault(require("crypto"));
const csp_header_1 = require("csp-header");
const psl = __importStar(require("psl"));
const constants_1 = require("./constants");
__export(require("./constants"));
function expressCspHeader(params) {
    return function (req, res, next) {
        if (!params) {
            next();
            return;
        }
        let { domainOptions } = params;
        let cspString = getCspString(req, res, params);
        cspString = applyNonce(req, cspString);
        cspString = applyAutoTld(req, cspString, domainOptions);
        setHeader(res, cspString, params);
        next();
    };
}
exports.expressCspHeader = expressCspHeader;
function getCspString(req, res, params) {
    let { directives, presets, reportUri } = params;
    let cspHeaderParams = {
        directives,
        presets,
        reportUri: typeof reportUri === 'function' ? reportUri(req, res) : reportUri
    };
    return csp_header_1.getCSP(cspHeaderParams);
}
function applyNonce(req, cspString) {
    if (cspString.includes(constants_1.NONCE)) {
        req.nonce = crypto_1.default.randomBytes(16).toString('base64');
        return cspString.replace(new RegExp(constants_1.NONCE, 'g'), csp_header_1.nonce(req.nonce));
    }
    return cspString;
}
function applyAutoTld(req, cspString, domainOptions) {
    if (cspString.includes(constants_1.TLD)) {
        let tld = parseDomain(req.hostname, domainOptions);
        if (!tld) {
            return cspString;
        }
        return cspString.replace(new RegExp(constants_1.TLD, 'g'), tld);
    }
    return cspString;
}
function parseDomain(hostname, domainOptions) {
    let customTlds = domainOptions === null || domainOptions === void 0 ? void 0 : domainOptions.customTlds;
    if (customTlds instanceof RegExp) {
        const tld = hostname.match(customTlds);
        if (tld !== null) {
            return tld[0].replace(/^\.+/, '');
        }
    }
    if (Array.isArray(customTlds)) {
        for (const tld of customTlds) {
            if (hostname.endsWith(`.${tld}`)) {
                return tld;
            }
        }
    }
    let domain = psl.parse(hostname);
    if (domain.error) {
        return null;
    }
    return domain.tld;
}
const CSP_HEADER = 'Content-Security-Policy';
const CSP_REPORT_ONLY_HEADER = 'Content-Security-Policy-Report-Only';
function setHeader(res, cspString, params) {
    let headerName = params.reportOnly ? CSP_REPORT_ONLY_HEADER : CSP_HEADER;
    res.set(headerName, cspString);
}
